### DISCUSSION TITLE
Custom Endpoint for storefront

### DISCUSSION THREAD

**Discussion by codewarsfx** [medusajs/medusa#3628] [Q&A]
Hello team,
So for my Usecase I'm trying to build a multi vendor market place with Medusa . to give it  multivendor abilities i extended  the Medusa core with the medusa Extender . For the storefront i'm using the NextJs storefront . I am trying to implement a feature on the storefront where i fetch product by StoreId ... to do this i created a custom endpoint on the  medusa server.  I do not intend to make any changes to the default storefront so i'm trying to closely mirror the data it expects.  To implement my custom endpoint i followed the same implementation of the /store/products route but the data being returned still differs from what is fetched on the next client. 

here's a sample response returned from my custom endpoint 

```
{
  "message": "success",
  "data": [
    {
      "id": "prod_01GVDDGX0WWTJMAQAH96BAPR4Q",
      "created_at": "2023-03-13T12:01:04.002Z",
      "updated_at": "2023-03-13T12:01:04.002Z",
      "deleted_at": null,
      "title": "f",
      "subtitle": "f",
      "description": null,
      "handle": "f",
      "is_giftcard": false,
      "status": "published",
      "thumbnail": "http://127.0.0.1:9001/medusa/334536258_1337820643704454_2036893078011829350_n-1678708863915.jpeg",
      "profile_id": "sp_01GVDCP26PB1H80D6WMKTW699P",
      "weight": null,
      "length": null,
      "height": null,
      "width": null,
      "hs_code": null,
      "origin_country": null,
      "mid_code": null,
      "material": "f",
      "collection_id": null,
      "type_id": null,
      "discountable": true,
      "external_id": null,
      "metadata": null,
      "store_id": "store_01GVDCWC50D58YEWF11H27550J",
      "variants": [
        {
          "id": "variant_01GVDDGX1TZ4Y08X31S8FN63J6",
          "created_at": "2023-03-13T12:01:04.002Z",
          "updated_at": "2023-03-15T17:42:28.573Z",
          "deleted_at": null,
          "title": "blu",
          "product_id": "prod_01GVDDGX0WWTJMAQAH96BAPR4Q",
          "sku": null,
          "barcode": null,
          "ean": null,
          "upc": null,
          "inventory_quantity": 2,
          "allow_backorder": false,
          "manage_inventory": true,
          "hs_code": null,
          "origin_country": null,
          "mid_code": null,
          "material": null,
          "weight": null,
          "length": null,
          "height": null,
          "width": null,
          "metadata": null,
          "options": [
            {
              "id": "optval_01GVDDGX1VYBRQ2KV3S1K84EEF",
              "created_at": "2023-03-13T12:01:04.002Z",
              "updated_at": "2023-03-13T12:01:04.002Z",
              "deleted_at": null,
              "value": "blu",
              "option_id": "opt_01GVDDGX14PR5MY72992HB9WFV",
              "variant_id": "variant_01GVDDGX1TZ4Y08X31S8FN63J6",
              "metadata": null
            }
          ],
          "prices": [
            {
              "id": "ma_01GVDDGX1Y3BCCH6PQXQWW0EAS",
              "created_at": "2023-03-13T12:01:04.002Z",
              "updated_at": "2023-03-13T12:01:04.002Z",
              "deleted_at": null,
              "currency_code": "usd",
              "amount": 200000,
              "min_quantity": null,
              "max_quantity": null,
              "price_list_id": null,
              "variant_id": "variant_01GVDDGX1TZ4Y08X31S8FN63J6",
              "region_id": null
            },
            {
              "id": "ma_01GVDDGX200Y9D0E9BMHA3D20J",
              "created_at": "2023-03-13T12:01:04.002Z",
              "updated_at": "2023-03-13T12:01:04.002Z",
              "deleted_at": null,
              "currency_code": "eur",
              "amount": 222200,
              "min_quantity": null,
              "max_quantity": null,
              "price_list_id": null,
              "variant_id": "variant_01GVDDGX1TZ4Y08X31S8FN63J6",
              "region_id": null
            }
          ]
        }
      ]
    }
  ]
}
```

which differs from what is being returned in the products  variable here

<img width="1440" alt="Screenshot 2023-03-29 at 13 00 23" src="https://user-images.githubusercontent.com/41517041/228528997-9f96306d-cb29-4b72-9d72-afae9074d96f.png">

my  plan is to make a call to my custom endpoint  here and replace the products variable with the data returned from my endpoint but this isn't possible if the data don't match up.... does this call from the medusa client 

```
const products= await
    medusaClient.products
    .list({
      is_giftcard: false,
      limit: 4,
      cart_id: cartId,
    })
```

do something more than call the /store/products endpoint.... am i missing something ???  any help will be appreciated 
thank you

**Comment by Simon-Bertrand** [comment]
> Hello team, So for my Usecase I'm trying to build a multi vendor market place with Medusa . to give it multivendor abilities i extended the Medusa core with the medusa Extender . For the storefront i'm using the NextJs storefront . I am trying to implement a feature on the storefront where i fetch product by StoreId ... to do this i created a custom endpoint on the medusa server. I do not intend to make any changes to the default storefront so i'm trying to closely mirror the data it expects. To implement my custom endpoint i followed the same implementation of the /store/products route but the data being returned still differs from what is fetched on the next client.
> 
> here's a sample response returned from my custom endpoint
> 
> ```
> {
>   "message": "success",
>   "data": [
>     {
>       "id": "prod_01GVDDGX0WWTJMAQAH96BAPR4Q",
>       "created_at": "2023-03-13T12:01:04.002Z",
>       "updated_at": "2023-03-13T12:01:04.002Z",
>       "deleted_at": null,
>       "title": "f",
>       "subtitle": "f",
>       "description": null,
>       "handle": "f",
>       "is_giftcard": false,
>       "status": "published",
>       "thumbnail": "http://127.0.0.1:9001/medusa/334536258_1337820643704454_2036893078011829350_n-1678708863915.jpeg",
>       "profile_id": "sp_01GVDCP26PB1H80D6WMKTW699P",
>       "weight": null,
>       "length": null,
>       "height": null,
>       "width": null,
>       "hs_code": null,
>       "origin_country": null,
>       "mid_code": null,
>       "material": "f",
>       "collection_id": null,
>       "type_id": null,
>       "discountable": true,
>       "external_id": null,
>       "metadata": null,
>       "store_id": "store_01GVDCWC50D58YEWF11H27550J",
>       "variants": [
>         {
>           "id": "variant_01GVDDGX1TZ4Y08X31S8FN63J6",
>           "created_at": "2023-03-13T12:01:04.002Z",
>           "updated_at": "2023-03-15T17:42:28.573Z",
>           "deleted_at": null,
>           "title": "blu",
>           "product_id": "prod_01GVDDGX0WWTJMAQAH96BAPR4Q",
>           "sku": null,
>           "barcode": null,
>           "ean": null,
>           "upc": null,
>           "inventory_quantity": 2,
>           "allow_backorder": false,
>           "manage_inventory": true,
>           "hs_code": null,
>           "origin_country": null,
>           "mid_code": null,
>           "material": null,
>           "weight": null,
>           "length": null,
>           "height": null,
>           "width": null,
>           "metadata": null,
>           "options": [
>             {
>               "id": "optval_01GVDDGX1VYBRQ2KV3S1K84EEF",
>               "created_at": "2023-03-13T12:01:04.002Z",
>               "updated_at": "2023-03-13T12:01:04.002Z",
>               "deleted_at": null,
>               "value": "blu",
>               "option_id": "opt_01GVDDGX14PR5MY72992HB9WFV",
>               "variant_id": "variant_01GVDDGX1TZ4Y08X31S8FN63J6",
>               "metadata": null
>             }
>           ],
>           "prices": [
>             {
>               "id": "ma_01GVDDGX1Y3BCCH6PQXQWW0EAS",
>               "created_at": "2023-03-13T12:01:04.002Z",
>               "updated_at": "2023-03-13T12:01:04.002Z",
>               "deleted_at": null,
>               "currency_code": "usd",
>               "amount": 200000,
>               "min_quantity": null,
>               "max_quantity": null,
>               "price_list_id": null,
>               "variant_id": "variant_01GVDDGX1TZ4Y08X31S8FN63J6",
>               "region_id": null
>             },
>             {
>               "id": "ma_01GVDDGX200Y9D0E9BMHA3D20J",
>               "created_at": "2023-03-13T12:01:04.002Z",
>               "updated_at": "2023-03-13T12:01:04.002Z",
>               "deleted_at": null,
>               "currency_code": "eur",
>               "amount": 222200,
>               "min_quantity": null,
>               "max_quantity": null,
>               "price_list_id": null,
>               "variant_id": "variant_01GVDDGX1TZ4Y08X31S8FN63J6",
>               "region_id": null
>             }
>           ]
>         }
>       ]
>     }
>   ]
> }
> ```
> 
> which differs from what is being returned in the products variable here
> 
> <img alt="Screenshot 2023-03-29 at 13 00 23" width="1440" src="https://user-images.githubusercontent.com/41517041/228528997-9f96306d-cb29-4b72-9d72-afae9074d96f.png">
> 
> my plan is to make a call to my custom endpoint here and replace the products variable with the data returned from my endpoint but this isn't possible if the data don't match up.... does this call from the medusa client
> 
> ```
> const products= await
>     medusaClient.products
>     .list({
>       is_giftcard: false,
>       limit: 4,
>       cart_id: cartId,
>     })
> ```
> 
> do something more than call the /store/products endpoint.... am i missing something ??? any help will be appreciated thank you

Hi,

Don't know if too late, but I could resolve the same issue as your using `medusaClient.client.request("GET", "/store/your/custom/api/endpoint")` function where medusaClient is imported from `import { medusaClient } from "@lib/config"`.

It is way better than instanciating the Medusa client React from medusa-react, where we would be using two different client libs

**Reply by feyiamujo1** [reply]
To call a custom endpoint on the storefront, you can call the custom endpoint using this (version 1.20) - 
`const { products, count } = await medusaClient.client.request(
      "GET",
      "/store/custom/sale-products-route",
      {
        pageParam: 0,
        queryParams: {
          ...queryParams,
        },
        countryCode,
      }
    )`


Ensure to test your endpoint's response before calling it from the storefront
